Fix input range init

Narretz opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- ~~[ ] Docs have been added / updated (for bug fixes / features)~~

**Other information**:

This PR fixes 1 bigger and 2 smaller issues:
- correctly initialize with interpolated min / max values

The interpolation directive only sets the actual element attribute value after
a digest passed. That means previously the min/max values on input range were
not set when the first $render happened, so the browser would not adjust
the input value according to min/max. This meant the range input and model would
not be initialzed as expected.

With this change, input range will set the actual element attribute value during its own
linking phase, as it is already available on the attrs argument passed to the link fn.

Fixes #14982
- correctly sets the range value when min > max. Chrome and IE don't set this correctly in all cases.
- removes interpolation from an ng-maxlength attribute in a test (interpolation in an attribute that expects an expression is discouraged)

Narretz committed:
fix(input[range]): correctly initialize with interpolated min / max values

The interpolation directive only sets the actual element attribute value after
a digest passed. That means previously, the min/max values on input range were
not set when the first $render happened, so the browser would not adjust
the input value according to min/max. This meant the range input and model would
not be initialzed as expected.

With this change, input range will set the actual element attribute value during its own
linking phase, as it is already available on the attrs argument passed to the link fn.

Fixes #14982

Narretz committed:
test(validators): remove bad practice from test

Narretz committed:
fix(input[range]): set correct value when max < min

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1647,7 +1647,8 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
       // IE11 doesn't set the el val correctly if the maxVal is greater than the element value
```
I just noticed this comment should say "less than" (instead of "greater than"), right?

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1647,7 +1647,8 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
         element.val(maxVal);
-        elVal = minVal;
+        // IE11 and Chrome don't set the value to the minVal when max < min
+        elVal = maxVal < minVal ? minVal : maxVal;
```
Shouldn't this be reflected to `element.val()` then? I would expect something like:

```js
if (maxVal < elVal) {
  elVal = Math.max(minVal, maxVal);
  element.val(elVal);
}
```

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1587,36 +1632,19 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
     if (supportsRange && maxAttrType === 'max') {
       var elVal = element.val();
-      // IE11 doesn't set the el val correctly if the maxVal is less than the element value
+      // IE11 doesn't set the el val correctly if the maxVal is greater than the element value
```
So change seems invalid.

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1563,11 +1565,54 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
+  }
+
+  function setInitialValueAndObserver(attrType, attrName, changeFn) {
+   var initialAttrValue;
```
Wrong indentation.

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1563,11 +1565,54 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
       val = parseFloat(val);
     }
-    minVal = isNumber(val) && !isNaN(val) ? val : undefined;
+    minVal = isNumber(val) && !isNaN(val) ? val : 0;
```
Why this change? Doesn't it break our "convention" that an invalid or undefined value results to no validation?
With this `minVal` can never be undefined.

Or is this trying to emulate the default browser behavior for browsers that don't natively support `input[range]`?

Narretz replied:
Good point. I assume that browsers that don't force valid values should not assume min = 0 and max = 100.

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -3245,22 +3270,21 @@ describe('input', function() {
-          expect(inputElm.val()).toBe('50');
-          expect(scope.value).toBe(50);
+          expect(inputElm.val()).toBe('60');
+          expect(scope.value).toBe(60);
```
So, we even had tests that verified the buggy behavior, nice :stuck_out_tongue:

Narretz replied:
Yeah, I don't know what I was thinking ...

gkalpak replied:
I LGTM'd that, so I don't know what I was thinking either :stuck_out_tongue:

Narretz commented:
@gkalpak Travis finally green, you may review this again :>

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1563,6 +1565,49 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
+
+    // This initalizes the min/max value, so that non-support browsers validate with the correct
+    // values during the initial $render
+    changeFn(initialAttrValue);
```
`initialAttrValue` will be undefined for `ngMin`/`ngMax`. Is this intended?

It is not clear why this function would do different things for `ngXyz` vs `xyz` attributes. If this is intended, you could add a short comment explaining what is going on.

Narretz replied:
At this point we really only need to set the element attribute if we have min / max and the browser supports range. I've removed the call to changeFn as its use isn't tested currently. 
The actual issue is that all validators that are based on observation if interpolation or ngX attributes only set their values after a digest. That means when the input renders these validators' valuea aren't defined. When the observers fire we will validate again with the correct values but this causes the model to be unset if the value is invalid. Imo this should still count as initialization so that shouldn't happen. The other thing is that we validate too often (once for every observer). 
I'll open an issue for part 1 of this.

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -3051,7 +3108,7 @@ describe('input', function() {
 
       it('should validate', function() {
-        var inputElm = helper.compileInput('<input type="range" ng-model="value" name="alias" ng-min="50" />');
+        var inputElm = helper.compileInput('<input type="range" ng-model="value" name="alias" ng-min="20" />');
```
Ooc, does this change make any difference?

Narretz replied:
Nope. I'll revert it.

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -3164,6 +3240,34 @@ describe('input', function() {
           expect(scope.form.alias.$error.max).toBeFalsy();
         });
 
+        it('should not assume a max val of 100 if the max attribute interpolates to a non-number', function() {
```
I don't see how this test illustrates this description. (But I am not sure if it is the description or the test or the implementation that needs to change :smiley:)

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1593,19 +1589,16 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
+      // interpolated attributes set the attribute value only after a digest, but we need the
+      // attribute value when the input is first rendered, so that the browser can adjust the
+      // input value based on the min/max value
+      element.attr(htmlAttrName, attr[htmlAttrName]);
```
So, if I understand correctly, when using `min`/`max` (with or without interpolation) the browser will set an initial value (if none is set) to `(max + min) / 2`. But if you use `ngMin`/`Max` the browser will ignore them and still assume the default `min`/`max` values (0 and 100 resp.) and set the initial value of the input to 50.
Is that correct?

Additionally, when using `ngMin`/`Max`, the browser will ignore their values and only allow values between the default 0 and 100.
This doesn't sound very useful... :confused:

Narretz replied:
Yes, you are correct. If you put it like this, it indeed doesn't sound very useful. It's all because ngMin/Max et. al. doesn't set the actual attribute on the element. I assume it's done for a reason, but I haven't investigated.

Narretz replied:
What if we simply document that you shouldn't use ngMin/ngMax with input range? Or not document them at all?

gkalpak replied:
Iirc, `ngMin/Max` will set `attrs.min/max`, but not the element's attribute or property. The reason is that we want to be able to have in-Angular validation, but not validation from the browser. This makes sense for other inputs, but not for `input[range]`, because it cannot have invalid values (in supporting browsers at least).

So, in a sense, the browser forces its restrictions on the user (they can't enter a value that is outside of the specified range). Because of that, I agree that `ngMin/Max` don't have anything to offer for `input[range]`.

I would add a little warning in the docs, though, saying that `ngMin/Max` will hav no effect on `input[range]`, because I expect many people will try to use them :smile:

Narretz replied:
Okay, then I'll remove ngMin/Max handling and docs, and add a note to the docs instead.

Narretz committed:
fix(input[range]): remove support for ngMin / ngMax

Since ng prefixed attributes do not set their corresponding element attribute, the range input would always have min = 0,
and max = 100 (in supported browsers), regardless of the value
in ngMin/ngMax. This is confusing and not very useful, so
it's better to not support them at all.

gkalpak commented:
LGTM

icorne commented:
Will this fix be back-ported to older angular versions?

petebacondarwin commented:
@icorne - which versions did you have in mind? We have no plans to backport it further back than 1.5.x

icorne commented:
I'm having trouble figuring out where this was merged into. I'm working on 1.5.x, so that's what I'm interested in. I've tried upgrading from 1.5.5 to 1.5.9 but I still get the issue.

petebacondarwin commented:
Ahah! Right. OK, it didn't make it into 1.5.9, since that was primarily a security release. It will be in 1.5.10.

gkalpak commented:
This is already included in 1.5.9 (2e7121b), but support for `input[range]` is opt-in on 1.5.x in order to avoid a breaking change (see 07b8761 and [the docs][1]).

[1]: https://code.angularjs.org/1.5.9/docs/api/ng/input/input%5Brange%5D

petebacondarwin commented:
Oops! I am so confused with all the rebasing we did last week that I forgot that we actually did put this in 1.5.9 :-) Thanks @gkalpak
